astronomer_starship/_af3/starship_compatability.py

```diff
@@ -0,0 +1,965 @@
+
+        # populate dag_version_id
+        dag_ids = {ti.get("dag_id") for ti in task_instances if ti.get("dag_id")}
+        for dag_id in dag_ids:
+            try:
+                self.update_dag_version_id(dag_id)
+                logger.info("Auto-populated dag_version_id for DAG: %s", dag_id)
+            except Exception as e:
+                logger.warning("Failed to auto-populate dag_version_id for %s: %s", dag_id, e)
+
+        return {"task_instances": task_instances}
+
+    def get_task_instance_history(self, dag_id: str, offset: int = 0, limit: int = 10):
+        from airflow.models import DagRun
+        from airflow.models.taskinstancehistory import TaskInstanceHistory
+        from sqlalchemy import desc
+        from sqlalchemy.orm import load_only
+
+        try:
+            sub_query = (
+                self.session.query(DagRun.run_id)
+                .filter(DagRun.dag_id == dag_id)
+                .order_by(desc(DagRun.start_date))
+                .limit(limit)
+            )
+            if offset:
+                sub_query = sub_query.offset(offset)
+            sub_query = sub_query.subquery()
+
+            results = (
+                self.session.query(TaskInstanceHistory)
+                .filter(TaskInstanceHistory.dag_id == dag_id)
+                .filter(TaskInstanceHistory.run_id.in_(sub_query))
+                .options(
+                    load_only(
+                        *[
+                            attr_desc["attr"]
+                            for attr, attr_desc in self.task_instance_attrs().items()
+                            if attr_desc["attr"] is not None and attr != "id"  # Exclude 'id' property
+                        ]
+                    )
+                )
+                .order_by(desc(TaskInstanceHistory.start_date))
+                .all()
+            )
+            return {
+                "task_instances": results_to_list_via_attrs(results, self.task_instance_attrs()),
+                "dag_run_count": self._get_dag_run_count(dag_id),
+            }
+        except Exception as e:
+            self.session.rollback()
+            raise e
+
+    def set_task_instance_history(self, task_instances: list):
+        task_instances = self.insert_directly("task_instance_history", task_instances)
+        return {"task_instances": task_instances}
+
+    def insert_directly(self, table_name, items):  # noqa: C901
+        import pickle
+        import uuid
+
+        from sqlalchemy import MetaData
+        from sqlalchemy.dialects.postgresql import insert
+        from sqlalchemy.exc import InvalidRequestError
+
+        if not items:
+            return []
+
+        for item in items:
+            # For task_instance/task_instance_history: generate new UUIDs
+            # to avoid PK conflicts with source data
+            if "id" in item:
+                if table_name in ["task_instance", "task_instance_history"]:
+                    item["id"] = str(uuid.uuid4())
```

Comment:
@akshaykumarsalunke I am wondering if we really need a new random UUID for task instances. 🤔 

Since those IDs are already random in the source. Wouldn't re-generation just have the same likelihood of conflicts? Maybe even worse because the source IDs were already unique within their environment.

I'd like to have a really good reason to re-generate those IDs in the target, **because this will make it much much harder to migrate task reschedules (less important), task instance notes and HITL detail** if we decide to add support for those in the future.

<img width="747" height="857" alt="Image" src="https://github.com/user-attachments/assets/64470b5f-eee9-4c00-a910-e6729337358b" />

Reply:
Done